### PR TITLE
Enhance GPSampler performance (other than introducing local search)

### PR DESCRIPTION
**Pull request by contramundum53**
## Motivation
We split PR #5274 into two. This PR introduces all improvements other than local search.

## Description of the changes
See #5274.

### CODE REVIEW

File: optuna/_gp/gp.py [hunk 1]
```diff
@@ -177,25 +180,31 @@ def loss_func(raw_params: np.ndarray) -> tuple[float, np.ndarray]:
             kernel_scale=torch.exp(raw_params_tensor[n_params]),
             noise_var=torch.exp(raw_params_tensor[n_params + 1]) + minimum_noise,
         )
+        if deterministic:
+            params.noise_var = torch.tensor(minimum_noise, dtype=torch.float64)
         loss = -marginal_log_likelihood(
             torch.from_numpy(X), torch.from_numpy(Y), torch.from_numpy(is_categorical), params
         ) - log_prior(params)
         loss.backward()  # type: ignore
         return loss.item(), raw_params_tensor.grad.detach().numpy()  # type: ignore
```
Comment (contramundum53) [inline optuna/_gp/gp.py:189]:
```suggestion
        if deterministic:
            assert raw_params_tensor.grad[n_params + 1] == 0
        return loss.item(), raw_params_tensor.grad.detach().numpy()  # type: ignore
```

File: optuna/_gp/prior.py [hunk 2]
```diff
@@ -20,13 +20,16 @@ def default_log_prior(kernel_params: "gp.KernelParamsTensor") -> "torch.Tensor":
...
         return (concentration - 1) * torch.log(x) - rate * x
 
     # NOTE(contramundum53): The parameters below were picked qualitatively.
     # TODO(contramundum53): Check whether these priors are appropriate.
     return (
-        gamma_log_prior(kernel_params.inverse_squared_lengthscales, 2, 0.5).sum()
+        -(
+            0.1 / kernel_params.inverse_squared_lengthscales
+            + 0.1 * kernel_params.inverse_squared_lengthscales
+        ).sum()
```
Comment (nabenabe0928) [inline optuna/_gp/prior.py:32]: What distribution's density function is it?

**Review by nabenabe0928 — COMMENTED** [review]
Thank you for separating the PR, I left some minor comments.

File: optuna/samplers/_gp/sampler.py [hunk 3]
```diff
@@ -105,6 +113,17 @@ def infer_relative_search_space(
 
         return search_space
 
+    def _optimize_acqf(
+        self,
+        acqf_params: "acqf.AcquisitionFunctionParams",
+    ) -> np.ndarray:
+        normalized_params, _acqf_val = optim_sample.optimize_acqf_sample(
```
[the remark this answers is not in the document]
Comment (contramundum53) [reply optuna/samplers/_gp/sampler.py:120]: This is for customization point (both for advanced users and for developers.) I overwrite this private function to perform benchmark across different optimization methods.

**Comment by contramundum53** [comment]
@nabenabe0928 I applied your comments. PTAL.

File: optuna/_gp/gp.py [hunk 4]
```diff
@@ -177,25 +180,33 @@ def loss_func(raw_params: np.ndarray) -> tuple[float, np.ndarray]:
...
             torch.from_numpy(X), torch.from_numpy(Y), torch.from_numpy(is_categorical), params
         ) - log_prior(params)
         loss.backward()  # type: ignore
+        # scipy.minimize requires all the gradients to be zero for termination.
+        assert not deterministic_objective or raw_params_tensor.grad[n_params + 1] == 0
         return loss.item(), raw_params_tensor.grad.detach().numpy()  # type: ignore
 
     # jac=True means loss_func returns the gradient for gradient descent.
-    res = so.minimize(loss_func, initial_raw_params, jac=True)
+    res = so.minimize(
```
Comment (nabenabe0928) [inline optuna/_gp/gp.py:194]: Could you add an inline comment why having `gtol` here is necessary?

File: optuna/_gp/gp.py [hunk 5]
```diff
@@ -177,25 +180,33 @@ def loss_func(raw_params: np.ndarray) -> tuple[float, np.ndarray]:
...
     raw_params_opt_tensor = torch.from_numpy(res.x)
 
-    return KernelParamsTensor(
+    res = KernelParamsTensor(
         inverse_squared_lengthscales=torch.exp(raw_params_opt_tensor[:n_params]),
         kernel_scale=torch.exp(raw_params_opt_tensor[n_params]),
         noise_var=torch.exp(raw_params_opt_tensor[n_params + 1]) + minimum_noise,
     )
+    if deterministic_objective:
+        res.noise_var = torch.tensor(minimum_noise, dtype=torch.float64)
```
Comment (nabenabe0928) [inline optuna/_gp/gp.py:208]: Can we do something like this so that we can keep frozen=True for KernelParamsTensor?
```suggestion
    noise_var = (
         torch.tensor(minimum_noise, dtype=torch.float64) if deterministic_objective
         else torch.exp(raw_params_opt_tensor[n_params + 1]) + minimum_noise
    )
    res = KernelParamsTensor(
        inverse_squared_lengthscales=torch.exp(raw_params_opt_tensor[:n_params]),
        kernel_scale=torch.exp(raw_params_opt_tensor[n_params]),
        noise_var=noise_var,
    )
```

File: optuna/_gp/prior.py [hunk 6]
```diff
@@ -20,13 +20,16 @@ def default_log_prior(kernel_params: "gp.KernelParamsTensor") -> "torch.Tensor":
     # Log of prior distribution of kernel parameters.
 
     def gamma_log_prior(x: "torch.Tensor", concentration: float, rate: float) -> "torch.Tensor":
-        # We omit the constant factor `rate ** concentration / factorial(concentration)`.
+        # We omit the constant factor `rate ** concentration / Gamma(concentration)`.
         return (concentration - 1) * torch.log(x) - rate * x
 
     # NOTE(contramundum53): The parameters below were picked qualitatively.
     # TODO(contramundum53): Check whether these priors are appropriate.
     return (
```
Comment (nabenabe0928) [inline optuna/_gp/prior.py:28]:
```suggestion
    # NOTE(contramundum53): Prior for inverse_squared_lengthscales was determined by heuristic.
    return (
```

File: optuna/samplers/_gp/sampler.py [hunk 7]
```diff
@@ -105,6 +113,17 @@ def infer_relative_search_space(
 
         return search_space
 
+    def _optimize_acqf(
+        self,
+        acqf_params: "acqf.AcquisitionFunctionParams",
+    ) -> np.ndarray:
+        normalized_params, _ = optim_sample.optimize_acqf_sample(
```
Comment (nabenabe0928) [inline optuna/samplers/_gp/sampler.py:120]: We can leave it for now, but let me add some comments because we should delete this method unless we get its strong demand from users.
```suggestion
        # This method is supposed to be overridden by advanced users or developers
        # to try different optimization strategies.
        # TODO(nabenabe0928): Remove this method once our optimization strategy is fixed.
        normalized_params, _ = optim_sample.optimize_acqf_sample(
```

File: optuna/_gp/gp.py [hunk 8]
```diff
@@ -177,25 +180,33 @@ def loss_func(raw_params: np.ndarray) -> tuple[float, np.ndarray]:
...
         ) - log_prior(params)
         loss.backward()  # type: ignore
+        # scipy.minimize requires all the gradients to be zero for termination.
+        assert not deterministic_objective or raw_params_tensor.grad[n_params + 1] == 0
         return loss.item(), raw_params_tensor.grad.detach().numpy()  # type: ignore
 
     # jac=True means loss_func returns the gradient for gradient descent.
-    res = so.minimize(loss_func, initial_raw_params, jac=True)
+    res = so.minimize(
+        loss_func, initial_raw_params, jac=True, method="l-bfgs-b", options={"gtol": gtol}
```
Comment (contramundum53) [inline optuna/_gp/gp.py:196]:
```suggestion
        # We need a high gtol value because the loss_func can have high numerically errors.
        loss_func, initial_raw_params, jac=True, method="l-bfgs-b", options={"gtol": gtol}
```

File: optuna/_gp/gp.py (hunk 4 above)
Comment (contramundum53) [reply optuna/_gp/gp.py:194]: Done.

File: optuna/_gp/gp.py (hunk 5 above)
Comment (contramundum53) [reply optuna/_gp/gp.py:209]: Personally keeping the `KernelParamsTensor` frozen doesn't particularly make sense, especially because what it has are `torch.Tensor`s whose gradients are designed to be rewritten. But I do not have a strong opinion.

File: optuna/_gp/prior.py (hunk 6 above)
Comment (contramundum53) [reply optuna/_gp/prior.py:28]: Well, every prior here are determined by heuristic, and that's no special about `inverse_squared_lengthscales`. I thought `The parameters below were picked qualitatively.` was talking about this.

File: optuna/samplers/_gp/sampler.py (hunk 7 above)
Comment (contramundum53) [reply optuna/samplers/_gp/sampler.py:120]: Do you have a strong reason that this function should be removed?
People have different budget requirements. Some people can put up with 1 hour optimizing their acquisition function, and others cannot stand 1 second. I don't think we can ever make a silver bullet optimization method that works for everyone.

File: optuna/_gp/prior.py [hunk 9]
```diff
@@ -20,13 +20,16 @@ def default_log_prior(kernel_params: "gp.KernelParamsTensor") -> "torch.Tensor":
     # Log of prior distribution of kernel parameters.
 
     def gamma_log_prior(x: "torch.Tensor", concentration: float, rate: float) -> "torch.Tensor":
-        # We omit the constant factor `rate ** concentration / factorial(concentration)`.
+        # We omit the constant factor `rate ** concentration / Gamma(concentration)`.
         return (concentration - 1) * torch.log(x) - rate * x
 
     # NOTE(contramundum53): The parameters below were picked qualitatively.
```
Comment (contramundum53) [inline optuna/_gp/prior.py:26]:
```suggestion
    # NOTE(contramundum53): The priors below were picked by heuristics.
```

File: optuna/_gp/prior.py (hunk 6 above)
Comment (contramundum53) [reply optuna/_gp/prior.py:28]: I changed the sentence to `The priors below were picked by heuristics.`

File: optuna/samplers/_gp/sampler.py (hunk 7 above)
Comment (contramundum53) [reply optuna/samplers/_gp/sampler.py:120]:
```suggestion
        # Advanced users can override this method to change the optimization algorithm.
        # However, we do not make any effort to keep backward compatibility between versions.
        normalized_params, _ = optim_sample.optimize_acqf_sample(
```
How about this comment?

File: optuna/_gp/gp.py (hunk 5 above)
Comment (contramundum53) [reply optuna/_gp/gp.py]: Done.

File: optuna/_gp/prior.py [hunk 10]
```diff
@@ -20,13 +20,16 @@ def default_log_prior(kernel_params: "gp.KernelParamsTensor") -> "torch.Tensor":
     # Log of prior distribution of kernel parameters.
 
     def gamma_log_prior(x: "torch.Tensor", concentration: float, rate: float) -> "torch.Tensor":
-        # We omit the constant factor `rate ** concentration / factorial(concentration)`.
+        # We omit the constant factor `rate ** concentration / Gamma(concentration)`.
         return (concentration - 1) * torch.log(x) - rate * x
 
-    # NOTE(contramundum53): The parameters below were picked qualitatively.
+    # NOTE(contramundum53): The priors below were picked by heuristics.
```
Comment (contramundum53) [inline optuna/_gp/prior.py:26]:
```suggestion
    # NOTE(contramundum53): The priors below (the functions and its params) were picked by heuristics.
```

File: optuna/_gp/gp.py [hunk 11]
```diff
@@ -175,27 +178,45 @@ def loss_func(raw_params: np.ndarray) -> tuple[float, np.ndarray]:
...
         loss.backward()  # type: ignore
+        # scipy.minimize requires all the gradients to be zero for termination.
+        raw_noise_var_grad = raw_params_tensor.grad[n_params + 1]  # type: ignore
+        assert not deterministic_objective or raw_noise_var_grad == 0
         return loss.item(), raw_params_tensor.grad.detach().numpy()  # type: ignore
 
     # jac=True means loss_func returns the gradient for gradient descent.
-    res = so.minimize(loss_func, initial_raw_params, jac=True)
+    res = so.minimize(
+        # We need a high gtol value because the loss_func can have high numerically errors.
```
Comment (nabenabe0928) [inline optuna/_gp/gp.py:198]:
```suggestion
        # Too small `gtol` causes instability in loss_func optimization.
```

File: optuna/_gp/gp.py (hunk 5 above)
Comment (nabenabe0928) [reply optuna/_gp/gp.py]: This comment was made by mistake, but so i will count on you if you would like to revert the change.

File: optuna/samplers/_gp/sampler.py (hunk 7 above)
Comment (nabenabe0928) [reply optuna/samplers/_gp/sampler.py:120]: Sounds great!

**Review by nabenabe0928 — APPROVED** [review]
Thank you for your work, LGTM!

File: optuna/_gp/gp.py [hunk 12]
```diff
@@ -164,7 +166,8 @@ def _fit_kernel_params(
             np.log(initial_kernel_params.inverse_squared_lengthscales.detach().numpy()),
             [
                 np.log(initial_kernel_params.kernel_scale.item()),
-                np.log(initial_kernel_params.noise_var.item() - minimum_noise),
+                # We add 0.01 * minimum_noise to initial noise_var to avoid instability.
+                np.log(initial_kernel_params.noise_var.item() - 0.99 * minimum_noise),
```
Comment (y0z) [inline optuna/_gp/gp.py:170]: Based on the comment above (i.e., "add 0.01 * minimum_noise to initial noise_var"), I feel this line should be like below.
```
np.log(initial_kernel_params.noise_var.item() + 0.01 * minimum_noise)
```

Comment (nabenabe0928) [reply optuna/_gp/gp.py:170]: I think this line is trying to transform the variable so that the search space is gonna be $(-\infty, \infty)$.

Comment (y0z) [reply optuna/_gp/gp.py:170]: Memo: The intention is below.
```
(initial_noise_var + 0.01 * minimum_noise) - minimum_noise = initial_var - 0.99 * minimum_noise.
```

**Review by y0z — APPROVED** [review]
LGTM.
